### PR TITLE
Move almost all functions' parameter db.Engine to context.Context

### PR DESCRIPTION
**Pull request by lunny**
As title.

### CODE REVIEW

File: models/issue.go [hunk 1]
```diff
@@ -597,7 +596,7 @@ func (issue *Issue) ReadBy(ctx context.Context, userID int64) error {
 		return err
 	}
 
-	return setIssueNotificationStatusReadIfUnread(db.GetEngine(db.DefaultContext), userID, issue.ID)
+	return setIssueNotificationStatusReadIfUnread(db.DefaultContext, userID, issue.ID)
```
Comment (KN4CK3R) [inline models/issue.go]:
```suggestion
	return setIssueNotificationStatusReadIfUnread(ctx, userID, issue.ID)
```
Should pass the context to `UpdateIssueUserByRead` too.

File: models/issue_label.go [hunk 2]
```diff
@@ -358,11 +358,12 @@ func GetLabelsInRepoByIDs(repoID int64, labelIDs []int64) ([]*Label, error) {
...
 }
 
-func getLabelsByRepoID(e db.Engine, repoID int64, sortType string, listOptions db.ListOptions) ([]*Label, error) {
+func getLabelsByRepoID(ctx context.Context, repoID int64, sortType string, listOptions db.ListOptions) ([]*Label, error) {
 	if repoID <= 0 {
 		return nil, ErrRepoLabelNotExist{0, repoID}
 	}
 	labels := make([]*Label, 0, 10)
+	e := db.GetEngine(ctx)
 	sess := e.Where("repo_id = ?", repoID)
```
Comment (KN4CK3R) [inline models/issue_label.go]:
```suggestion
	sess := db.GetEngine(ctx).Where("repo_id = ?", repoID)
```

File: models/project/board.go [hunk 3]
```diff
@@ -157,30 +156,30 @@ func deleteBoardByID(ctx context.Context, boardID int64) error {
 		return err
 	}
 
-	if err = board.removeIssues(e); err != nil {
+	if err = board.removeIssues(ctx); err != nil {
 		return err
 	}
 
-	if _, err := e.ID(board.ID).Delete(board); err != nil {
+	if _, err := db.GetEngine(ctx).ID(board.ID).NoAutoCondition().Delete(board); err != nil {
```
Comment (KN4CK3R) [inline models/project/board.go:163]: Why is this needed?

File: models/repo/fork.go [hunk 4]
```diff
@@ -10,16 +10,16 @@ import (
...
+	return repos, db.GetEngine(ctx).
 		Where("fork_id=?", forkID).
 		Find(&repos)
 }
 
 // GetRepositoriesByForkID returns all repositories with given fork ID.
 func GetRepositoriesByForkID(ctx context.Context, forkID int64) ([]*Repository, error) {
-	return getRepositoriesByForkID(db.GetEngine(ctx), forkID)
+	return getRepositoriesByForkID(ctx, forkID)
 }
```
Comment (KN4CK3R) [inline models/repo/fork.go]: There should be only the `GetRepositoriesByForkID` method.

File: models/repo/user_repo.go [hunk 5]
```diff
@@ -59,29 +59,29 @@ func CountUserRepositories(userID int64, private bool) int64 {
...
 func GetPublicRepositoryCount(u *user_model.User) (int64, error) {
-	return getPublicRepositoryCount(db.GetEngine(db.DefaultContext), u)
+	return getPublicRepositoryCount(db.DefaultContext, u)
 }
 
 // GetPrivateRepositoryCount returns the total number of private repositories of user.
 func GetPrivateRepositoryCount(u *user_model.User) (int64, error) {
-	return getPrivateRepositoryCount(db.GetEngine(db.DefaultContext), u)
+	return getPrivateRepositoryCount(db.DefaultContext, u)
 }
```
Comment (KN4CK3R) [inline models/repo/user_repo.go]: Keep only the exported methods.

File: models/project/board.go (hunk 3 above)
Comment (lunny) [reply models/project/board.go:163]: `NoAutoCondition` means only use `ID` become the delete condition but not read any field from `board` here. It has performance improvement.

File: models/issue_assignees.go [hunk 6]
```diff
@@ -25,20 +25,15 @@ func init() {
...
-func (issue *Issue) LoadAssignees() error {
-	return issue.loadAssignees(db.GetEngine(db.DefaultContext))
-}
-
-// This loads all assignees of an issue
-func (issue *Issue) loadAssignees(e db.Engine) (err error) {
+func (issue *Issue) LoadAssignees(ctx context.Context) (err error) {
 	// Reset maybe preexisting assignees
 	issue.Assignees = []*user_model.User{}
+	issue.Assignee = nil
```
Comment (KN4CK3R) [inline models/issue_assignees.go:31]: 👍

File: models/issue_assignees.go [hunk 7]
```diff
@@ -147,34 +116,38 @@ func toggleIssueAssignee(ctx context.Context, issue *Issue, doer *user_model.Use
...
+	i := 0
+	for ; i < len(issue.Assignees); i++ {
 		if issue.Assignees[i].ID == assigneeID {
+			found = true
 			break
 		}
 	}
+	if !found {
+		return
+	}
```
Comment (KN4CK3R) [inline models/issue_assignees.go]: Is that correct? If the user was not assigned the insert code later cannot be run.

File: services/issue/assignee.go [hunk 8]
```diff
@@ -45,6 +47,7 @@ func DeleteNotPassedAssignee(issue *models.Issue, doer *user_model.User, assigne
 
 // ToggleAssignee changes a user between assigned and not assigned for this issue, and make issue comment for it.
 func ToggleAssignee(issue *models.Issue, doer *user_model.User, assigneeID int64) (removed bool, comment *models.Comment, err error) {
+	fmt.Println("11-----")
```
Comment (KN4CK3R) [inline services/issue/assignee.go]:
```suggestion
```

File: models/issue_assignees.go (hunk 7 above)
Comment (lunny) [reply models/issue_assignees.go]: Fixed

File: models/action_list.go [hunk 9]
```diff
@@ -85,7 +86,7 @@ func (actions ActionList) loadRepoOwner(e db.Engine, userMap map[int64]*user_mod
 		}
 		repoOwner, ok := userMap[action.Repo.OwnerID]
 		if !ok {
-			repoOwner, err = user_model.GetUserByID(action.Repo.OwnerID)
+			repoOwner, err = user_model.GetUserByIDCtx(ctx, action.Repo.OwnerID)
```
Comment (6543) [inline models/action_list.go:89]: bugfix - backporting?

File: models/commit_status.go [hunk 10]
```diff
@@ -49,21 +49,21 @@ func init() {
...
 			repoID, sha)
 	case setting.Database.UseMySQL:
-		_, err = e.Exec("INSERT INTO `commit_status_index` (repo_id, sha, max_index) "+
+		_, err = db.Exec(ctx, "INSERT INTO `commit_status_index` (repo_id, sha, max_index) "+
 			"VALUES (?,?,1) ON DUPLICATE KEY UPDATE max_index = max_index+1",
 			repoID, sha)
 	case setting.Database.UseMSSQL:
 		// https://weblogs.sqlteam.com/dang/2009/01/31/upsert-race-condition-with-merge/
-		_, err = e.Exec("MERGE `commit_status_index` WITH (HOLDLOCK) as target "+
+		_, err = db.Exec(ctx, "MERGE `commit_status_index` WITH (HOLDLOCK) as target "+
```
Comment (6543) [inline models/commit_status.go:66]:
```suggestion
		_, err = db.GetEngine(ctx).Exec("MERGE `commit_status_index` WITH (HOLDLOCK) as target "+
```
why not this :point_up: 

what are the benefits to recreate the interface (+context added) of engine with the db package?

File: models/repo/mirror.go [hunk 11]
```diff
@@ -56,21 +54,16 @@ func (m *Mirror) BeforeInsert() {
 	}
 }
 
-// AfterLoad is invoked from XORM after setting the values of all fields of this object.
-func (m *Mirror) AfterLoad(session *xorm.Session) {
-	if m == nil {
-		return
+// GetRepository returns the repository.
+func (m *Mirror) GetRepository() *Repository {
```
Comment (6543) [inline models/repo/mirror.go:58]: can we rename it LoadRepository() as with other loadAtributes ?

File: models/repo/pushmirror.go [hunk 12]
```diff
@@ -35,21 +33,16 @@ func init() {
 	db.RegisterModel(new(PushMirror))
 }
 
-// AfterLoad is invoked from XORM after setting the values of all fields of this object.
-func (m *PushMirror) AfterLoad(session *xorm.Session) {
-	if m == nil {
-		return
+// GetRepository returns the path of the repository.
+func (m *PushMirror) GetRepository() *Repository {
```
Comment (6543) [inline models/repo/pushmirror.go:37]: same ...

File: services/issue/assignee_test.go [hunk 13]
```diff
@@ -20,21 +21,24 @@ func TestDeleteNotPassedAssignee(t *testing.T) {
...
 
 	// Clean everyone
 	err = DeleteNotPassedAssignee(issue, user1, []*user_model.User{})
 	assert.NoError(t, err)
+	assert.EqualValues(t, 0, len(issue.Assignees))
 
 	// Check they're gone
-	assignees, err := models.GetAssigneesByIssue(issue)
+	err = issue.LoadAssignees(db.DefaultContext)
 	assert.NoError(t, err)
```
Comment (6543) [inline services/issue/assignee_test.go]:
```suggestion
	assert.NoError(t, issue.LoadAssignees(db.DefaultContext))
```
nit ...

File: services/issue/assignee.go [hunk 14]
```diff
@@ -21,9 +21,10 @@ import (
 // DeleteNotPassedAssignee deletes all assignees who aren't passed via the "assignees" array
 func DeleteNotPassedAssignee(issue *models.Issue, doer *user_model.User, assignees []*user_model.User) (err error) {
 	var found bool
+	oriAssignes := make([]*user_model.User, len(issue.Assignees))
+	_ = copy(oriAssignes, issue.Assignees)
```
Comment (6543) [inline services/issue/assignee.go:25]: unrelated bugfix/enhancement?

Comment (lunny) [reply services/issue/assignee.go:25]: Bugfix. Because issue.Assignees will be changed in Togglexxxx method.

File: models/action_list.go (hunk 9 above)
Comment (lunny) [reply models/action_list.go:89]: Right, I will send a new PR to backport it.

File: models/commit_status.go (hunk 10 above)
Comment (lunny) [reply models/commit_status.go:66]: It's shorter. :)

File: models/repo/mirror.go (hunk 11 above)
Comment (lunny) [reply models/repo/mirror.go:58]: Looks like we can't do that. Because the interface of `RemoteMirrorer` needs a method like that.
```go
type RemoteMirrorer interface {
	GetRepository() *Repository
	GetRemoteName() string
}
```

File: models/repo/pushmirror.go (hunk 12 above)
Comment (lunny) [reply models/repo/pushmirror.go:37]: As above.

File: models/action_list.go (hunk 9 above)
Comment (lunny) [reply models/action_list.go:89]: Looks like there is no this bug in v1.16

File: models/repo/mirror.go (hunk 11 above)
Comment (6543) [reply models/repo/mirror.go:58]: haha ....

**Review by 6543 — APPROVED** [review]

File: routers/web/repo/setting.go [hunk 15]
```diff
@@ -1214,6 +1214,7 @@ func selectPushMirrorByForm(form *forms.RepoSettingForm, repo *repo_model.Reposi
 	}
 
 	for _, m := range pushMirrors {
+		m.Repo = repo
 		if m.ID == id {
 			return m, nil
 		}
```
Comment (wxiaoguang) [inline routers/web/repo/setting.go]: Only small nit here. Since it's newly written code, it's strange that some m.Repo is assigned but not used. If the code will be changed in future (eg: `return` => `break`), the loop will leave some `m.Repo=nil` items in the slice (not friendly for future refactorings)

Changing or not, either is fine.

```suggestion
	for _, m := range pushMirrors {
		if m.ID == id {
			m.Repo = repo
			return m, nil
		}
```

**Comment by wxiaoguang** [comment]
Generally L-G-T-M. But I do not understand why some PRs have rebase+force-push after the reviews have started.

In my mine, rebase+force-push only do harm to the project, it makes the review more difficult, and even worse it may lead to code lost.

**Comment by lunny** [comment]
> Generally L-G-T-M. But I do not understand why some PRs have rebase + force-push after the reviews have started.
> 
> In my mind, rebase + force-push only do harm to the project, it makes the review more difficult, and even worse it may lead to code lost.

For me, that's because I'm also reviewing myself PR and found some code could be changed. Since Gitea asks to merge PR after updating, I have a habit to update the PR when possible to reduce the conflicts.

**Comment by wxiaoguang** [comment]
But rebase doens't help about conflicts. A normal merge is enough for all cases, and the commit history is clear.

**Review by wxiaoguang — APPROVED** [review]

**Comment by 6543** [comment]
Please resolv conflict ;)

**Comment by lunny** [comment]
make L-G-T-M work.
